src/app-layer-ftp.c

```diff
@@ -228,31 +272,121 @@ static int FTPParseRequest(Flow *f, void *ftp_state,
     /* toserver stream */
     state->direction = 0;
 
+    int direction = STREAM_TOSERVER;
     while (FTPGetLine(state) >= 0) {
         FTPParseRequestCommand(state,
                                state->current_line, state->current_line_len);
-        if (state->command == FTP_COMMAND_PORT) {
-            if (state->current_line_len > state->port_line_size) {
-                ptmp = SCRealloc(state->port_line, state->current_line_len);
-                if (ptmp == NULL) {
-                    SCFree(state->port_line);
-                    state->port_line = NULL;
-                    state->port_line_size = 0;
-                    return 0;
+        switch (state->command) {
+            case FTP_COMMAND_PORT:
+                if (state->current_line_len > state->port_line_size) {
+                    ptmp = SCRealloc(state->port_line, state->current_line_len);
+                    if (ptmp == NULL) {
+                        SCFree(state->port_line);
+                        state->port_line = NULL;
+                        state->port_line_size = 0;
+                        return 0;
+                    }
+                    state->port_line = ptmp;
+
+                    state->port_line_size = state->current_line_len;
                 }
-                state->port_line = ptmp;
-
-                state->port_line_size = state->current_line_len;
-            }
-            memcpy(state->port_line, state->current_line,
-                   state->current_line_len);
-            state->port_line_len = state->current_line_len;
+                memcpy(state->port_line, state->current_line,
+                        state->current_line_len);
+                state->port_line_len = state->current_line_len;
+                break;
+            case FTP_COMMAND_RETR:
+                /* change direction (default to server) so expectation will handle
+                 * the correct message when expectation will match.
+                 */ 
+                direction = STREAM_TOCLIENT;
+                // fallthrough
+            case FTP_COMMAND_STOR:
+                {
```

Comment:
can you add a comment that if we get here 'current_line_len' is at least >=5? This is the type of code that looks suspicious as it doesn't check things itself.

I would also not be opposed to actually checking the side here, purely as a defensive measure so future changes to FTPGetLine/FTPParseRequestCommand don't cause unintended issues

Reply:
*checking the size